crates/parry3d/examples/custom_dispatcher.rs

```diff
@@ -0,0 +1,327 @@
+    }
+
+    fn compute_local_bounding_sphere(&self) -> BoundingSphere {
+        self.0.compute_local_bounding_sphere()
+    }
+
+    fn clone_dyn(&self) -> Box<dyn Shape> {
+        Box::new(Self(self.0))
+    }
+
+    fn scale_dyn(&self, scale: &Vector<f32>, num_subdivisions: u32) -> Option<Box<dyn Shape>> {
+        Some(self.0.scale_dyn(scale, num_subdivisions)?)
+    }
+
+    fn mass_properties(&self, density: f32) -> MassProperties {
+        self.0.mass_properties(density)
+    }
+
+    fn shape_type(&self) -> ShapeType {
+        self.0.shape_type()
+    }
+
+    fn as_typed_shape(&self) -> TypedShape {
+        self.0.as_typed_shape()
+    }
+
+    fn ccd_thickness(&self) -> f32 {
+        self.0.ccd_thickness()
+    }
+
+    fn ccd_angular_thickness(&self) -> f32 {
+        self.0.ccd_angular_thickness()
+    }
+}
+
+pub struct CustomBallDispatcher;
+
+impl QueryDispatcher for CustomBallDispatcher {
+    fn intersection_test(
+        &self,
+        pos12: &Isometry<f32>,
+        g1: &dyn Shape,
+        g2: &dyn Shape,
+    ) -> Result<bool, Unsupported> {
+        let (ball1, ball2) = (
+            g1.downcast_ref::<CustomBall>(),
+            g2.downcast_ref::<CustomBall>(),
+        );
+
+        match (ball1, ball2) {
+            (Some(ball1), Some(ball2)) => {
+                let p12 = Point::from(pos12.translation.vector);
+                return Ok(parry3d::query::details::intersection_test_ball_ball(
+                    &p12, &ball1.0, &ball2.0,
+                ));
+            }
+            (Some(ball1), None) => {
+                let dispatcher = DefaultQueryDispatcher;
+                dispatcher.intersection_test(pos12, &ball1.0, g2)
+            }
+            (None, Some(ball2)) => {
+                let dispatcher = DefaultQueryDispatcher;
+                dispatcher.intersection_test(pos12, g1, &ball2.0)
+            }
+            _ => {
+                let dispatcher = DefaultQueryDispatcher;
+                dispatcher.intersection_test(pos12, g1, g2)
+            }
+        }
+    }
+
+    fn distance(
+        &self,
+        pos12: &Isometry<f32>,
+        g1: &dyn Shape,
+        g2: &dyn Shape,
+    ) -> Result<f32, Unsupported> {
+        let (ball1, ball2) = (
+            g1.downcast_ref::<CustomBall>(),
+            g2.downcast_ref::<CustomBall>(),
+        );
+
+        match (ball1, ball2) {
+            (Some(ball1), Some(ball2)) => {
+                let p2 = Point::from(pos12.translation.vector);
+                return Ok(parry3d::query::details::distance_ball_ball(
+                    &ball1.0, &p2, &ball2.0,
+                ));
+            }
+            (Some(ball1), None) => {
+                let dispatcher = DefaultQueryDispatcher;
+                dispatcher.distance(pos12, &ball1.0, g2)
+            }
+            (None, Some(ball2)) => {
+                let dispatcher = DefaultQueryDispatcher;
+                dispatcher.distance(pos12, g1, &ball2.0)
+            }
```

Comment:
These branches are not necessary for the example, we should replace them by a comment saying "implement the algorithm for each shape type combination you want to support."